In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
file_path = Path("Resources/lending_data.csv")
lending_df = pd.read_csv(file_path)
# Review the DataFrame
lending_df.head()
# YOUR CODE HERE!

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y = lending_df["loan_status"]
# Separate the X variable, the features
# YOUR CODE HERE!
X = lending_df.drop(columns=["loan_status"])
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [4]:
# Review the y variable Series
# YOUR CODE HERE!
y

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64

In [5]:
# Review the X variable DataFrame
# YOUR CODE HERE!
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
# YOUR CODE HERE!
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1,stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
classifier = LogisticRegression(solver='lbfgs',random_state=1)
# Fit the model using training data
# YOUR CODE HERE!
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
# YOUR CODE HERE!
prediction = classifier.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!
print(f"balanced accuracy score is: {balanced_accuracy_score(y_test, prediction)}")

balanced accuracy score is: 0.9442676901753825


In [11]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
c_matrix = confusion_matrix(y_test, prediction)
c_matrix

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [12]:
# Print the classification report for the model
# YOUR CODE HERE!
target_names = ["Low-Risk loan", "High-Risk loan"]
class_report = classification_report(y_test, prediction,target_names=target_names, digits = 4)
print(class_report)

                precision    recall  f1-score   support

 Low-Risk loan     0.9964    0.9957    0.9961     18759
High-Risk loan     0.8746    0.8928    0.8836       625

      accuracy                         0.9924     19384
     macro avg     0.9355    0.9443    0.9398     19384
  weighted avg     0.9925    0.9924    0.9925     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Regarding the logistic regression analysis, there was higher precision to detect the low and high risk. The overall accuracy detected was of 0.9924 and an F1-score of 0.9925. Showing strong indication of model performance in predicting both healthy and high-risk loans.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [19]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
oversample = RandomOverSampler(random_state=1)
# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!
X_resampled, y_resampled = oversample.fit_resample(X_train, y_train)

In [20]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!
y_resampled.value_counts()


0    56277
1    56277
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [21]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
classifier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
# YOUR CODE HERE!
classifier.fit(X_resampled, y_resampled)
# Make a prediction using the testing data
# YOUR CODE HERE!
pred_resampled = classifier.predict(X_resampled)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [23]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
score = balanced_accuracy_score(y_resampled, pred_resampled)
print(f"Balanced accuracy score of model is: {score:.3f}")

Balanced accuracy score of model is: 0.994


In [25]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
c_matrix2 = confusion_matrix(y_resampled, pred_resampled)
c_matrix2

array([[55945,   332],
       [  323, 55954]], dtype=int64)

In [26]:
# Print the classification report for the model
# YOUR CODE HERE!
target_names = ["Low-Risk loan", "High-Risk loan"]
resampled_report = classification_report(y_resampled, pred_resampled,target_names=target_names, digits = 4)
print(resampled_report)

                precision    recall  f1-score   support

 Low-Risk loan     0.9943    0.9941    0.9942     56277
High-Risk loan     0.9941    0.9943    0.9942     56277

      accuracy                         0.9942    112554
     macro avg     0.9942    0.9942    0.9942    112554
  weighted avg     0.9942    0.9942    0.9942    112554



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** RESAMPLING THE DATA HAVE INCREASED MODEL PREDICTION WITH SLIGHT DECREASE IN THE PRECISION, RECALL AND F1-SCORE. THE OVERALL BALANCE ACCURACY SCORE WAS MADE 0.994 SHOWING THE STRONG ACCURACY IN PREDICTION OF HEALTHY AND HIGH-RISK LOANS